onnxruntime/core/session/inference_session.cc

```diff
@@ -869,19 +865,31 @@ common::Status InferenceSession::TransformGraph(onnxruntime::Graph& graph,
                                                 const onnxruntime::GraphTransformerManager& graph_transformer_mgr,
                                                 const ExecutionProviders& providers,
                                                 KernelRegistryManager& kernel_registry_manager,
-                                                const InsertCastTransformer& insert_cast_transformer,
                                                 SessionState& session_state,
                                                 bool saving_model_in_ort_format) {
   // The transformer order:
   // 1. run level 1 optimizations. these only use ONNX operators.
-  // 2. partition nodes based on EP capabilities. EPs may fuse nodes during this process.
-  // 3. run all optimizations. level 2 and 3 optimizations use contrib ops.
-  // 4. insert cast nodes
-  // 5. insert copy nodes
+  // 2. ensure potential QDQ node units have unique DQ nodes (required transformer).
```

Comment:
should this run before level 1 transformers or before partitioning?
if before level 1 transformers, we can assume the expected DQ state (no DQ with multiple consumers) earlier on.
however, if before partitioning (which is where this condition is most important), we have less chance of something else modifying the graph and messing up the expected DQ state.
or it could be in both places, but maybe that's excessive.

Reply:
I think the safest is as early as possible in level 1 as that will fix changes made by other level 1 transformers as well as changes made by layout transform (which happens during partitioning) now that we (once #15117 is checked in) run the L1 transformers as part of that process. Important detail is we keep looping through the transformers in a level until they stop changing the graph or we hit the max iterations.

Reply:
hm, I see the benefits of including it in level 1. however, it would then be a "mandatory level 1 transformer", which seems like more of a special case (and unprecedented) than having it be a "required transformer". should it be an error if the user tries to exclude it, as they may do for other level 1 transformers?

Reply:
it could be an error. disabling transformers is rare (and currently only possible via the python API as it was added for ORT format model conversion) though, and wouldn't the check the QDQ NodeUnit is valid catch any issues if they mattered?

What's the reason for running it after level 1? I would have expected it would be better to get the graph into a known 'good' state earlier rather than later so the general assumption that DQ nodes have a single consumer can be made across all the optimizers and EPs.

I have a slight concern about it not being part of level 1 though as someone could add a new optimizer that results in DQ nodes with multiple consumers (which happened recently) and we may not catch it unless a test is a) affected by this new optimizer and b) hits an EP that works with QDQ NodeUnit instances. at least if it's in level 1 those two optimizers will fight to the death until the max iterations for the transformers is hit, and at the issue will be obvious at caught at development time.

Reply:
one other case to consider is when the user requests an optimization level of ORT_DISABLE_ALL to bypass even level 1 optimizations.

taking a step back, this DQ duplication is not so much an optimization as a required graph transformation. the level1, level2, level3 transformers are optimizations, and optional, right? so the DQ duplication does not seem to fit cleanly into one of the existing optimization groups.

> What's the reason for running it after level 1?

just to fix things up once after the level 1 transformers complete. ideally it runs at the point between other graph modifications and processing of QDQ node groups.

> I would have expected it would be better to get the graph into a known 'good' state earlier rather than later so the general assumption that DQ nodes have a single consumer can be made across all the optimizers and EPs.

I think this is a useful assumption to be able to make. we could also run it once before the level 1 graph transformers.

> I have a slight concern about it not being part of level 1 though as someone could add a new optimizer that results in DQ nodes with multiple consumers

it is a concern I have too. if there are transformer tests which test the new level 1 transformer's interaction with other ones via InferenceSession::Initialize(), the tests should also include the DQ duplication along with other level 1 transformers. [TransformerTester](https://github.com/microsoft/onnxruntime/blob/710d095124cc7fc74570c5b9f5257ded0ec43fdb/onnxruntime/test/optimizer/graph_transform_test_builder.cc#L45) does this. from the other end, I also added tests for EnsureUniqueDQForNodeUnit which test interaction with level 1 transformers. in these kinds of tests, I think there shouldn't be much difference between having DQ duplication as a separate required transformer or another level 1 transformer (as a required transformer, it would be even harder to disable). it's another matter for tests which test a transformer in isolation.

Reply:
went with running it both before level1 transformers and with level1 transformers